include/swift/AST/Decl.h

```diff
@@ -573,6 +573,9 @@ class alignas(1 << DeclAlignInBits) Decl {
     unsigned : NumTypeDeclBits;
 
     unsigned Recursive : 1;
+
+    /// Whether or not this declaration is currently being type-checked.
+    unsigned BeingTypeChecked : 1;
   };
   enum { NumAssociatedTypeDeclBits = NumTypeDeclBits + 1 };
```

Comment:
I wonder if it's even necessary.  I couldn't find uses for the Recursive flag, but I was worried I would be stepping on toes overwriting it.  Still, it seems like it can be merged.